teps/NNNN-catalog-testing.md

```diff
@@ -0,0 +1,495 @@
+information to express the idea and why it was not acceptable.
+-->
+
+### Options for what versions of Tekton Pipelines we test against
+
+We can run against all versions of Tekton Pipelines that a Task indicates it’s compatible
+with, or we can select a few to run against, for example:
+
+1. Run against all versions of Tekton Pipelines that a Task indicates it is compatible with,
+  starting from the minimum version the most recent version released
+2. Run against the minimum version and the most recent version released only
+3. Run against a moving window of releases, e.g. the most recent X releases
+
+We have proposed testing against only 2 releases on every PR (as well as the
+nightly release in periodic tests) to use the minimum amount of infrastructure
+required to be reasonably sure that the Task works.
+
+## Infrastructure
+
+We evolve
+[the existing pre-apply-task-hook.sh and pre-apply-taskrun-hook.sh](https://github.com/tektoncd/catalog/blob/43f52adb4dce0b7f7effb9e7f18e6c2d504c6e4b/DEVELOPMENT.md#end-to-end-testing)
+by:
+
+1. Implementing them with Tasks and a Pipeline that glues them together
+2. Providing preprovisioned configuration via parameters to
+   [the Task's test Pipeline](#verified-task-test-requirements).
+  
+We will iteratively provide more and more infrastructure for Tekton Tasks
+to test against. At a minimum we will provide:
+
+* An image registry (an instance of docker-registry running in the same namespace) that can be pushed to
+* A GCP project and ServiceAccount that can create and interact with GCS buckets, and GCR images
+* A kubeconfig configured to use a GKE cluster that can be deployed to and the required ServiceAccount
+
+We will then expand this infrastructure, additionally providing:
+
+* A GitHub repository and/or project with credentials that can be used to interact with it (e.g. opening PRs, updating checks)
+* A Slack workspace and user credentials so that a Task can post comments to Slack
+* We will deal with other requests on a case by case basis
+
+_We will work with the CDF such that the bill will be covered by the CDF and all infrastructure
+can be equally accessed by all OWNERS of the catalog repo._
+
+### Pipeline design
+
+The Pipeline that invokes these test Pipelines will consist of 3 Pipelines:
+
+1. A Pipeline that determines which Tasks need to be tested and runs the second Pipeline for each.
+  a. For pull requests, this will be only the Tasks that have changed
+  b. For nightly runs, this will be all Tasks
+2. A Pipeline which tests an individual Task by:
+  1. Doing the fast cheap checks first, in parallel:
+    a. Lint (including yamllint, conftest ….)
+    b. Verify requirements, including specifying the min version of Tekton Pipelines
+  2. Determine the the versions of Tekton Pipelines to test against
+  3. For each version to test this Task against, start the third Pipeline
+3. A Pipeline which tests an individual Task against a version of Tekton Pipelines
```

Comment:
I think requiring our pipelines to comprise of tasks that are already in the catalog would be less confusing for beginners. Also, it would reduce redundancy and improve the quality of our pipelines. It promotes smaller prs and tests will give an accurate indication on what is actually failing. I.E If the pipeline is assembled incorrectly, is the task failing because is there an edge case we did not account for. I feel like having a smorgasbord of tekton resources could lead to catastrophic failures.